source/common/http/http2/codec_impl.cc

```diff
@@ -283,17 +291,64 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
       stream->headers_->addViaMove(std::move(key), std::move(stream->cookies_));
     }
 
-    if (frame->headers.cat == NGHTTP2_HCAT_REQUEST || frame->headers.cat == NGHTTP2_HCAT_RESPONSE) {
+    if (frame->headers.cat == NGHTTP2_HCAT_REQUEST && stream->headers_->Expect() &&
+        0 == StringUtil::caseInsensitiveCompare(stream->headers_->Expect()->value().c_str(),
+                                                Headers::get().ExpectValues._100Continue.c_str())) {
+      // Deal with expect: 100-continue here since higher layers are never going to do anything
+      // other than say to continue so that we can respond before request complete if necessary.
+      std::vector<nghttp2_nv> final_headers;
+      StreamImpl::buildHeaders(final_headers, *CONTINUE_HEADER);
+      int rc = nghttp2_submit_headers(session_, 0, stream->stream_id_, nullptr, &final_headers[0],
+                                      final_headers.size(), nullptr);
+      ASSERT(rc == 0);
+      UNREFERENCED_PARAMETER(rc);
+
+      stream->headers_->removeExpect();
+    }
+
+    switch (frame->headers.cat) {
+    case NGHTTP2_HCAT_RESPONSE: {
+      uint64_t response_code = Http::Utility::getResponseStatus(*stream->headers_);
+      if (CodeUtility::is1xx(response_code)) {
+        stream->waiting_for_final_headers_ = true;
+      }
+
+      // Fall through.
+    }
+
+    case NGHTTP2_HCAT_REQUEST: {
       stream->decoder_->decodeHeaders(std::move(stream->headers_), stream->remote_end_stream_);
-    } else {
-      ASSERT(frame->headers.cat == NGHTTP2_HCAT_HEADERS);
-      ASSERT(stream->remote_end_stream_);
+      break;
+    }
 
-      // It's possible that we are waiting to send a deferred reset, so only raise trailers if local
-      // is not complete.
+    case NGHTTP2_HCAT_HEADERS: {
+      // It's possible that we are waiting to send a deferred reset, so only raise headers/trailers
+      // if local is not complete.
       if (!stream->deferred_reset_.valid()) {
-        stream->decoder_->decodeTrailers(std::move(stream->headers_));
+        if (!stream->waiting_for_final_headers_) {
+          ASSERT(stream->remote_end_stream_);
+          stream->decoder_->decodeTrailers(std::move(stream->headers_));
+        } else {
+          ASSERT(!nghttp2_session_check_server_session(session_));
+          stream->waiting_for_final_headers_ = false;
+
+          // This can only happen in the client case in a response, when we received a 1xx to
+          // start out with. In this case, raise as headers. nghttp2 message checking guarantees
+          // proper flow here.
+          // TODO(mattklein123): Higher layers don't currently deal with a double decodeHeaders()
+          // call and will probably crash. We do this for testing the server case. We correctly
+          // handle the server case. In the future, if needed, we can properly handle 1xx in higher
```

Comment:
Not clear what is meant by the server case here - how does the server receive a 100 response, isn't it the one generating the 100 response?